General combat calculation improvements for Avatars

NassirAlawar opened the pull request:
Too much changed to go into detail here. I have summarized changes in each commit and my code is packed full of comments and links to references and research.

Super short summary: 
Updated battleutils and summon.lua with equations for pDIF, fSTR, Hit Rate, Level Correction.
I have updated the base weapon computation for avatars.
I have updated how crit rate from dex works.
Level correction is not computed for newer zones/content since the release of SoA.
Some minor bug fixes.

Note: I have not updated the BP damage equations and plan to tackle those in the future, likely one avatar per PR or maybe even one BP per PR.

Note: I have not added the Damage Limit+ trait or "Physical Damage +%" to the computations as I do not believe the project supports them yet. I have noted how they were and where they need to be added once they are implemented.

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

NassirAlawar committed:
Crit damage modifiers are already caluclated in battleutils so it was being double counted here

NassirAlawar committed:
I have updated the hit rate calculations to match current known equations. There is honestly too much to list in this commit but I have heavily commented the code with links to where I pulled the data from and I will try to summarize below.

I have modified the return value of GetDexCritBonus because the crit rate can be negative.
I have updated the equations found in GetHitRateEx(CBattleEntity* PAttacker, CBattleEntity* PDefender, uint8 attackNumber, int8 offsetAccuracy)
I have updated the equations found in GetDexCritBonus(CBattleEntity* PAttacker, CBattleEntity* PDefender)
I have updated the equations found in GetDamageRatio(CBattleEntity* PAttacker, CBattleEntity* PDefender, bool isCritical, float bonusAttPercent)
I have updated the equations found in GetFSTR(CBattleEntity* PAttacker, CBattleEntity* PDefender, uint8 SlotID)

I have collected some data and compared against retail using level 1 mobs and a D1 weapon as well as shiva for avatar tests, the numbers for melee hits are quite accurate now (base stats still deviate from retail).

Level correct is now disabled in all newer zones and content.
Level correction for monsters and avatars caps at a delta of 38.
Hit Rate equations have been updated to match known current formulas.
Hit Rate cap now takes into account weapon type and offhand vs main.
Most weapon types and pets will now cap at 99% hit rate up from 95%.
Crit bonus from dDEX can be negative and caps at +-15.
Avatar fSTR has no cap.
Avatars get +8% to crit damage.
Avatar pDIF does not have a crit cap, just a hard cap of 4.25 before randomization.
pDIF caps take into account weapon type.
Crit attack bonus is now computed after randomization as it is a flat % increase to final damage.

NassirAlawar committed:
In 2014 pet base damage was updated, I do not have references to data for other pets but celestial avatars weapon damage is approximately equal to (1+level) now.
I have not tested carbuncle so the 0.9 is currently a guess based on the fact that the weapon damage was 10% less previously and carbuncle is still weaker than the celestials.
Happy to get numbers for carbuncle if needed.
Cait is the same as carbuncle, they suck equally.

NassirAlawar committed:
Merge branch 'release' of https://github.com/project-topaz/topaz into fix/avatars

NassirAlawar committed:
Typo with or

NassirAlawar committed:
This mimics the changes done in battleutils for DexCritRate, Hit Rate, fSTR, pDIF and Level Correction.
I have also removed the cap from the summoning skill bonus, it was set to 200 but retail is well over 250 now.
My SMN skill set has 695 total skill and I know I can break 700.
I have moved target:WakeUp to after the fTP damage modifier instead of before just in case.

As a note Blood Pact damage is still not accurate with this change but this ensures that the basic computations match retail.
It seems fTP values are somewhat incorrect as well as how fTP can cary over between hits which is confirmed on some BPs.
I plan to work towards fixing Blood Pact equations in another PR but I feel like this is large enough and focused on base damage calculations and hit rate.

NassirAlawar committed:
Fixing build check failures

zach2good commented on scripts/globals/summon.lua:
```diff
@@ -10,112 +10,278 @@ function getSummoningSkillOverCap(avatar)
-    local hitdmg = 0
+    hitrateSubsequent = utils.clamp(hitrateSubsequent, minHitRate, maxHitRate)
+    hitrateFirst = utils.clamp(hitrateFirst, minHitRate, maxHitRate)
+
+
```
No need for double line breaks here

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -2229,14 +2229,84 @@ namespace battleutils
+            //Grasshopper Broth / Noisy Grasshopper Broth / Mole Broth / Lively Mole Broth / Blood Broth / Clear Blood Broth / Antica Broth / Fragrant Antica Broth
+
+            int32 maxHitRate = 99;
+            auto targ_weapon = dynamic_cast<CItemWeapon*>(PAttacker->m_Weapons[SLOT_MAIN]);
```
This cast can still produce `nullptr`, which will get used by `targ_weapon->isTwoHanded();` later, you can change that to:
`bool isTwoHanded = targ_weapon && targ_weapon->isTwoHanded();` to make it safe

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -2331,99 +2443,213 @@ namespace battleutils
 
-        cRatio = std::clamp<float>(cRatio, 0, ratioCap);
+        float qRatio = wRatio;
+        float upperLimit = 0.0;
+        float lowerLimit = 0.0;
```
Floats should be marked as `0.0f` or else they'll be created as doubles and then converted. I'm sure the compiler probably handles this, but better to be clear

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -2331,99 +2443,213 @@ namespace battleutils
+        //https://www.bluegartr.com/threads/114636-Monster-Avatar-Pet-damage
+        //Monster pDIF = Avatar pDIF = Pet pDIF
+
+        auto targ_weapon = dynamic_cast<CItemWeapon*>(PAttacker->m_Weapons[SLOT_MAIN]);
```
Same here, you're going to be using `targ_weapon` later and there isn't a guarantee that it'll be populated, so either bail all of those checks if this is nullptr or put a nullptr check inline with each of those down below

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -2331,99 +2443,213 @@ namespace battleutils
+            //Avatars get the Crit. Atk. Bonus II trait for a +8% crit damage increase
+            if(attackerType == TYPE_PET)
+            {
+                CPetEntity* petEntity = dynamic_cast<CPetEntity*>(PAttacker);
+                if(petEntity)
```
Can be changed to:
```cpp
if (CPetEntity* petEntity = dynamic_cast<CPetEntity*>(PAttacker))
{
    ...
```

zach2good commented on src/map/utils/battleutils.cpp:
```diff
@@ -2487,6 +2713,25 @@ namespace battleutils
+
+            if(attackerType == TYPE_PET)
+            {
+                CPetEntity* petEntity = dynamic_cast<CPetEntity*>(PAttacker);
+                isAvatar = petEntity->getPetType() == PETTYPE_AVATAR;
```
petEntity can be nullptr here
```cpp
if (CPetEntity* petEntity = dynamic_cast<CPetEntity*>(PAttacker))
{
    isAvatar = petEntity->getPetType() == PETTYPE_AVATAR;
}
```

zach2good commented on src/map/utils/petutils.cpp:
```diff
@@ -1608,13 +1608,17 @@ namespace petutils
+            if (PetID == PETID_CARBUNCLE || PetID == PETID_CAIT_SITH)
             {
-                ((CItemWeapon*)PPet->m_Weapons[SLOT_MAIN])->setDamage((uint16)(floor(PPet->GetMLevel() * 0.67f)));
+                weaponDamage = (uint16)(floor(PPet->GetMLevel() * 0.9f));
```
Use `static_cast` instead of C-Style casts please

zach2good reviewed (commented):
I can't really speak for the maths content of the PR, this sort of balancing and stats stuff is _way_ beyond what I understand. Thanks for all the links to sources and explanations! They should help out people with a better understanding of whats going on here.

Lua all looks nice and tidy.

I have a couple of C++ dings:
General nullptr checking (because paranoia is good in core)
https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#breakbeforebraces-allman
https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#spacebeforeparens-controlstatements
https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#casting---static_cast-over-c-style

And maybe an excuse to introduce a new rule:
In lua you're leaving a space after starting a comment: `-- Comment`
Can you do the same with C++: `// Comment` instead of `//Comment`

Otherwise, everything looks spot-on! 👍 

EDIT: Added the comment thing as a rule
https://github.com/project-topaz/topaz/blob/release/CONTRIBUTING.md#general-code-guidlines-all-languages

TeoTwawki commented:
> And maybe an excuse to introduce a new rule:
In lua you're leaving a space after starting a comment: -- Comment
Can you do the same with C++: // Comment instead of //Comment

new? that was always a rule. I know coz it was my rule before topaz happened. :stuck_out_tongue:

ibm2431 commented on scripts/globals/summon.lua:
```diff
@@ -10,112 +10,278 @@ function getSummoningSkillOverCap(avatar)
+
+    local zoneId = avatar:getZone():getID()
+
+    local shouldApplyLevelCorrection = (zoneId < 256) and not (zoneId == 183)
```
[Should throw in the source for this](https://www.bg-wiki.com/bg/PDIF#Level_Correction_Function_.28cRatio.29), because I was gearing up to fight it. 😉 

Although we should probably be setting up and using a new zone flag instead of checking hardcoded zone IDs.

genoxd replied:
lol sure I'll add it, I left out the link because I'd been linking to pDIF so often. I thought I was safe!
I would agree with a zone flag, the first place I looked was the DB to see if there was any flag per expansion and or anything else I could use. Luckily they're all clumped together except for Legion A

NassirAlawar commented:
Sure, I'll fix these today 👍 
Just to be clear on the thread name: "General combat calculation improvements for Avatars"
This will impact players as well, I started with the summon.lua but then that drifted into the general combat equations for the melee hits.

ibm2431 commented on scripts/globals/summon.lua:
```diff
@@ -10,112 +10,278 @@ function getSummoningSkillOverCap(avatar)
+
+            --Final pDif is qRatio randomized with a 1-1.05 multiplier
+            local pDif = qRatio * (1 + (math.random() * 0.05))
+            --Avatars have Crit Atk Bonus II for +8% crit dmg
+            if isCrit then
+                pDif = pDif * 1.08
+            end
```
Let's give/use a mod for the Crit Bonus via trait~!

genoxd replied:
Sure, I'll look into doing that

ibm2431 commented on scripts/globals/summon.lua:
```diff
@@ -10,112 +10,278 @@ function getSummoningSkillOverCap(avatar)
+
+        -- If the first hit kills the mob other hits are not computed in retail
+        -- This can easily be seen by using BPs on low level mobs where the first hit kills them
+        while numHitsProcessed < numHitsLanded and finaldmg < targetHP do
```
Problem with the `finaldmg < targetHP` check is that `AvatarPhysicalMove` is called _before_ defense adjustments for the defender `AvatarFinalAdjustments`, which can cause these cycles to break out early under the assumption the mob will be dead.

genoxd replied:
Oh interesting okay, maybe down the road we need a final display damage and an internal total damage. Good catch 👍

ibm2431 commented on scripts/globals/summon.lua:
```diff
@@ -10,112 +10,278 @@ function getSummoningSkillOverCap(avatar)
-    local pdif = math.random(pdifMin * 1000, pdifMax * 1000) / 1000
-    if math.random() < critrate then
-        pdif = math.min(pdif + 1, 4.2)
+        if finaldmg > 0 then
+            target:wakeUp()
+        end
```
I know this was here before, but waking up targets is now baked into lua_baseentity:takeDamage(); you can remove it from the raw damage calculation step.

NassirAlawar committed:
Removed the targetHP check so it does not cause issues with the final phase of damage computation.
Removed target:wakeUp since it is covered elsewhere
Added reference for the level correction

NassirAlawar committed:
Added a reference to level correction in accuracy calculation.
Changed all c style casting to static_cast.
Added null checks for targ_weapon and petEntity.
Cleaned up if statements to match style guide.
Cleaned up comments to match style guide.

NassirAlawar committed:
Changed c style cast to static_cast.
Updated comments to match style guide

NassirAlawar committed:
Accidentally erased a ( in an if statement.
Fixed an error in the comments stating Physical Damange +% increases final pDIF, it increases the final pDIF cap meaning the upper limit for pDIF.

Retested to make sure nothings crashing after clean up

NassirAlawar committed:
Merge branch 'release' of https://github.com/project-topaz/topaz into fix/avatars

# Conflicts:
#	src/map/utils/battleutils.cpp

NassirAlawar committed:
Removed unused variable from merge

NassirAlawar commented:
I believe all the noted issues have been fixed and I have merged and retested

NassirAlawar commented:
Just realized I forgot to implement the request to give avatars crit att bonus from db and get rid of the hard coding. Will try to get that done in a day or two

NassirAlawar committed:
Removed hardcoded +8 crit damage and instead applied to to all avatars as a crit damage modifier.
Updated summon.lua to use the modifier as well.

I have retested player melee, avatar melee and avatar BPs to confirm everything still works.

NassirAlawar committed:
Merge branch 'release' of https://github.com/project-topaz/topaz into fix/avatars

NassirAlawar commented:
Changed the crit bonus to a modifier and added the modifier logic to summon.lua as well

I have retested and confirmed I see no issues. As always, let me know if there's anything else 👍

zach2good commented:
We were talking about this PR yesterday; Avatars will eventually need a full rewrite, but this lays out a well-researched framework for when that happens 👍

NassirAlawar commented:
Let me know if there's anything I can do to help regarding avatars or if there's anything you'd prefer to have prioritized. My next plan was to move onto blood pact equations including modifiers. I have been collecting some Pet: DEX+/STR+ gear for math stuff on retail.

zach2good commented:
I honestly know nothing about combat balancing and rely on the others for sanity checking, but if you continue testing things and documenting your sources, we'll have all of that available when someone "fixes" avatars 👍
